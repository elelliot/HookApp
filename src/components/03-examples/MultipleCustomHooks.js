import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effects.css';


export const MultipleCustomHooks = () => {


    const { counter,increment } = useCounter(1);

    const { loading, data } = useFetch( `https://www.breakingbadapi.com/api/quotes/${ counter }` );
    // console.log( state );

    //!!data (si lo que se evalua es true, partiendo de null= null, !null= true !!null= false)  
    //&& ejecuta y retorna lo siguiente (data[0])
    const { author , quote } = !!data && data[0];
    

    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr />

            {//if
                loading
                ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                :
                    //else
                    (
                        <blockquote className= "blockquote text-right">
                            <p className="mb-0"> { quote } </p>
                            <footer className="blockquote-footer">{ author }</footer>
                        </blockquote>
                    )
            }
            {/* Por alguna raz√≥n el Snapshot no captura el onClick, pero si lo pongo dentro del else como quiero, ni
            siquiera me detecta el boton en el Snapshot..whyyyy*/}
                <button 
                    className="btn btn-primary"
                    onClick ={ increment }    
                >
                    Next Quote
                </button>

        </div>
    )
}
