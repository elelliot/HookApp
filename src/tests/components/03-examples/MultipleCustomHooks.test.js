import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

//El mock es básicamente un Fake deployment del hook en este archivo, los argumentos nosotros los inicialisamos
jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");




describe('Pruebas en <MultipleCustomHooks /> ', () => {
    

    //Ya que no se va cambiar el valor, lo ponemos aquí merengue
    
    /*No sirve sin el beforeEach, pero al profe le sale, perhaps por las versiones de react y enzyme diferentes?, idk
      */
    beforeEach( () => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
    })



    test('Debe mostrarse correctamente', () => {

        //Default data para useFetch ya que estamos simulando el hook en uso con el mock
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect( wrapper ).toMatchSnapshot();
        
    });


    test('Debe mostrar la info', () => {
       

        //Inicialisamos con info que nosotros creamos
        useFetch.mockReturnValue({
            data:[{
                author: 'ElElliot',
                quote: 'This mofo wildin'
            }],
            loading:false,
            error:null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        //Si hay info no debería haber alert
        expect( wrapper.find('.alert').exists() ).toBe( false );
        //Que el texto de la clase .mb-0 (es un parrafo) sea lo que yo puse en el quote del mockReturn
        expect( wrapper.find('.mb-0').text().trim() ).toBe( 'This mofo wildin' );
        //Que el autor sea el correcto, ahora por tag en vez de seleccionar por clase
        expect( wrapper.find('footer').text().trim() ).toBe( 'ElElliot' );


        // console.log( wrapper.html() );
    });
    
    
})
