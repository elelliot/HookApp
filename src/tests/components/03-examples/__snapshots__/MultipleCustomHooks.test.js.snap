// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks />  Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</div>
`;
